pennylane/qnode.py

```diff
@@ -200,6 +201,10 @@ def __init__(
         self.expansion_strategy = expansion_strategy
         self.max_expansion = max_expansion
 
+        # The NullQubit doesn't handle caching.
+        if isinstance(device, NullQubit):
+            cache = False
+

```

Comment:
So caching wasn't working because of this line: https://github.com/PennyLaneAI/pennylane/blob/a8dacc8c9be7b0ed546feaba6ebf0fa460354991/pennylane/interfaces/execution.py#L588

`lambda x: qml.math.shape(x)[0]` is run on the result of the execution, and that fails when `x` is None.  

If we make it so that `dev.execute` returns `[0.0]` instead, then we caching works and we don't need this line.

Reply:
Thank you. I will probably set it to return [0.0].
It is not a priority to make the null qubit work with caching, but I definitely agree that we shouldn't add more checks here if possible. :slightly_smiling_face: